File: Client/src/com/kasania/net/Connection.java
package com.kasania.net;

import java.io.IOException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetSocketAddress;
import java.nio.ByteBuffer;
import java.nio.channels.ClosedByInterruptException;
import java.nio.channels.DatagramChannel;
import java.nio.channels.SocketChannel;
import java.util.Objects;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.FutureTask;

public final class Connection {

    private SocketChannel socketChannel;
    private DatagramSocket imageSocket;
    private DatagramSocket audioSocket;

    private ExecutorService executorService;
    private FutureTask<?> task;


    private ExecutorService executorService2;
    private FutureTask<?> task2;


    private ExecutorService executorService3;
    private FutureTask<?> task3;

    private Sender sender;
    private Receiver receiver;

    public static int VideoPort;
    public static int AudioPort;

    public void connect(String address, int port, int imagePort, int audioPort){

        VideoPort = imagePort;
        AudioPort = audioPort;

        new Thread(() ->{

            try {
                socketChannel = SocketChannel.open();
                socketChannel.connect(new InetSocketAddress(address,port));

                imageSocket = new DatagramSocket(new InetSocketAddress(VideoPort));
                audioSocket = new DatagramSocket(new InetSocketAddress(AudioPort));

                sender = new Sender();
                receiver = new Receiver();

                receiver.addReader(this::read);
                receiver.addImageReader(this::readImage);
                receiver.addAudioReader(this::readAudio);
                sender.addSender(this::write);

                executorService = Executors.newSingleThreadExecutor();
                task = new FutureTask<>(receiver::readData);
                executorService.execute(task);

                executorService2 = Executors.newSingleThreadExecutor();
                task2 = new FutureTask<>(receiver::readImageData);
                executorService2.execute(task2);

                executorService3 = Executors.newSingleThreadExecutor();
                task3 = new FutureTask<>(receiver::readAudioData);
                executorService3.execute(task3);

            } catch (IOException e) {
                e.printStackTrace();
                disconnect();
            }
        }).start();
    }

    ByteBuffer read(){

        ByteBuffer size = ByteBuffer.allocate(Integer.BYTES);
        ByteBuffer data = null;

        try {
            socketChannel.read(size);
            size.flip();
            int sz = size.getInt();
            data = ByteBuffer.allocate(sz);
            while(data.hasRemaining()){
                socketChannel.read(data);
            }
        }
        catch(ClosedByInterruptException ex){
            //Disconnected
        }
        catch (IOException e) {
            e.printStackTrace();
            disconnect();
        }
        Objects.requireNonNull(data).flip();
        return data;
    }

    ByteBuffer readImage(){

        ByteBuffer data = ByteBuffer.allocate(16384);

        try {
            byte[] buffer = new byte[16384];
            DatagramPacket datagramPacket = new DatagramPacket(buffer, buffer.length);
            imageSocket.receive(datagramPacket);
            data.put(buffer);
        } catch (IOException e) {
            e.printStackTrace();
        }
        return data.flip();
    }

    ByteBuffer readAudio(){

        ByteBuffer data = ByteBuffer.allocate(3528 + Integer.BYTES);

        try {

            byte[] buffer = new byte[3528 + Integer.BYTES];
            DatagramPacket datagramPacket = new DatagramPacket(buffer, 3528 + Integer.BYTES);
            audioSocket.receive(datagramPacket);
            data.put(buffer);
        } catch (IOException e) {
            e.printStackTrace();
        }
        return data.flip();
    }

    void write(ByteBuffer byteBuffer){
        try {
            if(socketChannel.isConnected()){
                ByteBuffer data = ByteBuffer.allocate(Integer.BYTES + byteBuffer.limit());
                data.putInt(byteBuffer.limit());
                data.put(byteBuffer);
                data.flip();
                socketChannel.write(data);
            }
        }
        catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void disconnect(){

        try {
            receiver.shutdown();
            task.cancel(true);
            executorService.shutdown();

            sender.addSender((_closed)->{});
            socketChannel.close();
        }
        catch (IOException e) {
            e.printStackTrace();
        }
    }

}

File: Server/src/com/kasania/server/connection/Connection.kt
package com.kasania.server.connection

import com.kasania.server.DataType
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetSocketAddress
import java.nio.ByteBuffer
import java.nio.channels.DatagramChannel
import java.nio.channels.SocketChannel
import java.util.concurrent.atomic.AtomicBoolean

open class Connection(val socketChannel: SocketChannel, private var type: Type) {

    enum class Type{
        DESKTOP,MOBILE,PENDING
    }

    private val datagramChannel = DatagramSocket()

    val syncDone : AtomicBoolean = AtomicBoolean(false)

    fun changeType(type: Type){
        this.type = type
    }
    fun getType(): Type {
        return this.type
    }
    private fun makePacket(dataType: DataType, src:Int, data: ByteBuffer): ByteBuffer? {

        val packagedData = ByteBuffer.allocate(Character.BYTES + Integer.BYTES + data.limit())
        packagedData.putChar(dataType.code)
        packagedData.putInt(src)
        packagedData.put(data)
        packagedData.flip()
        return packagedData

    }

    fun send(dataType: DataType, src:Int, data: ByteBuffer){
        makePacket(dataType,src,data)?.let { send(it) }
    }

    private fun send(data: ByteBuffer){
        if (socketChannel.isConnected) {
            val pendingData = ByteBuffer.allocate(Integer.BYTES + data.limit())
            pendingData.putInt(data.limit())
            pendingData.put(data)
            pendingData.flip()
            socketChannel.write(pendingData)
        }
    }

    fun sendImage(imagePort :Int, data: ByteBuffer){
        datagramChannel.send(DatagramPacket(data.array(), data.limit(), InetSocketAddress((socketChannel.remoteAddress as InetSocketAddress).address, imagePort)))
    }

    fun sendAudio(audioPort :Int, data: ByteBuffer){
        datagramChannel.send(DatagramPacket(data.array(), data.limit(), InetSocketAddress((socketChannel.remoteAddress as InetSocketAddress).address, audioPort)))
    }

    fun sendImage2(imagePort :Int, data: ByteBuffer){
        val packagedData = ByteBuffer.allocate(Character.BYTES + Integer.BYTES + data.limit())
        packagedData.putChar(DataType.IMAGE.code)
        packagedData.put(data)
        packagedData.flip()
        send(packagedData)
    }

    fun sendAudio2(audioPort :Int, data: ByteBuffer){
        val packagedData = ByteBuffer.allocate(Character.BYTES + Integer.BYTES + data.limit())
        packagedData.putChar(DataType.IMAGE.code)
        packagedData.put(data)
        packagedData.flip()
        send(packagedData)
    }


    fun syncDone(port:Int) {
        val buffer = ByteBuffer.allocate(4)
        buffer.putInt(port)
        buffer.flip()
        send(DataType.SYNCDone, 0, buffer)
    }
}